File: BallClock.py
import time

class BallClock:
    def __init__(self, ballCount):
        self.ballCount = ballCount
        self.halfDays = 0
        self.main = self.createBallSet()
        # createQueue is for comparisons
        self.createQueue = self.createBallSet()
        self.minutes = []
        self.fiveMinutes = []
        self.hours = []

    def startClock(self):
        startTime = time.time()
        while True:
            self.addMinute(self.main.pop(0))
            if self.isBallBackToStart() == True:
                break
        endTime = time.time()

        message = "%s balls cycle after %s days." %(self.ballCount, self.halfDays / 2)
        print message

        print "Completed in %s milliseconds (%s seconds)" %(((endTime-startTime) * 1000), endTime-startTime)

        return message

    # creates the balls in an array for both main and the comparison row/track: createQueue
    def createBallSet(self):
        queue = []
        for ball_number in range(1, self.ballCount + 1):
            queue.append(ball_number)
        return queue

    # this function checks to see if main has returned to it's initial state created in createQueue
    def isBallBackToStart(self):
        if len(self.main) != self.ballCount:
            return False
        for i in range(self.ballCount):
            if self.main[i] != self.createQueue[i]:
                return False
        return True

    # this function adds a ball to the minute row/track
    def addMinute(self, ball):
        if len(self.minutes) < 4:
            self.minutes.append(ball)
        else:
            self.clearMinutesRow()
            self.addFiveMinutes(ball)

    # this function adds a ball to the five minute row/track
    def addFiveMinutes(self, ball):
        if len(self.fiveMinutes) < 11:
            self.fiveMinutes.append(ball)
        else:
            self.clearFiveMinutesRow()
            self.addHour(ball)

    # this function adds a ball to the hour row/track
    def addHour(self, ball):
        if len(self.hours) < 11:
            self.hours.append(ball)
        else:
            self.clearHoursRow()
            self.main.append(ball)
            self.halfDays += 1

    # this function clears the minutes row/track
    def clearMinutesRow(self):
        while len(self.minutes) > 0:
            self.main.append(self.minutes.pop())

        return self.minutes

    # this function clears the five minute row/track
    def clearFiveMinutesRow(self):
        while len(self.fiveMinutes) > 0:
            self.main.append(self.fiveMinutes.pop())

        return self.fiveMinutes

    # this function clears the hours row/track
    def clearHoursRow(self):
        while len(self.hours) > 0:
            self.main.append(self.hours.pop())

        return self.hours

# BallClockMinutes creates a ball clock with x number of balls and a minute_count to track the time elapsed
class BallClockMinutes(BallClock):
    def __init__(self, ballCount, minute_count):
        BallClock.__init__(self, ballCount)
        self.minute_count = minute_count

    def startClock(self):
        for i in range(1, self.minute_count + 1):
            self.addMinute(self.main.pop(0))
        json_var = {"Min": self.minutes, "FiveMin": self.fiveMinutes, "Hour": self.hours, "Main": self.main}
        print json_var




a = BallClock(45)
a.startClock()

b = BallClockMinutes(30, 325)
b.startClock()

File: BallClock.js
class BallClock {
    constructor(ballsCount) {
        this.ballsCount = ballsCount;
        this.halfDay = 0;
        this.main = this.initBallArray();
        this.initQueue = this.initBallArray();
        this.mins = [];
        this.fiveMinutes = [];
        this.hoursAmount = [];
    }

    startClock() {
        // # also I needed to make sure that it was added one to the minute track, changed to while true
        var startTime = performance.now()
        while (true) {
            this.addOneMin(this.main.shift());
            if (this.checkReachedStart() === true) {
                break;
            }
        }
        var endTime = performance.now()

        var message = this.ballsCount + " balls cycle after " + (this.halfDay / 2) + " days.";
        console.log(message);

        var timeToRun = "Completed in " +  (endTime - startTime) + " milliseconds (" + ((endTime-startTime)/1000) + " seconds)";
        console.log(timeToRun);

        return message;
    }

    checkReachedStart() {
        if (this.main.length !== this.ballsCount) {
            return false;
        }

        for (var i = 0; i < this.ballsCount; i++) {
            if (this.main[i] !== this.initQueue[i]) {
                return false;
            }
        }
        return true;
    }

    // # the error existed here: I was not doing <= so it was off by one
    initBallArray() {
        var ballArray = [];
        for (var i = 1; i <= this.ballsCount; i++) {
            ballArray.push(i);
        }
        return ballArray;
    }

    addOneMin(ball) {
        if (this.mins.length < 4) {
            this.mins.push(ball)
        } else {
            this.clearMins();
            this.addFiveMin(ball);
        }
    }

    addFiveMin(ball) {
        if (this.fiveMinutes.length < 11) {
            this.fiveMinutes.push(ball);
        } else {
            this.clearFiveMins();
            this.addOneHour(ball);
        }
    }

    addOneHour(ball) {
        if (this.hoursAmount.length < 11) {
            this.hoursAmount.push(ball);
        } else {
            this.clearHours();
            this.main.push(ball);
            this.halfDay = this.halfDay + 1;
        }
    }

    clearMins() {
        while (this.mins.length > 0) {
            this.main.push(this.mins.pop());
        }
    }

    clearFiveMins() {
        while (this.fiveMinutes.length > 0) {
            this.main.push(this.fiveMinutes.pop());
        }
    }

    clearHours() {
        while (this.hoursAmount.length > 0) {
            this.main.push(this.hoursAmount.pop());
        }
    }
}

class BallClockMins extends BallClock {
    constructor(ballsCount, minCount) {
        super(ballsCount);
        this.minCount = minCount;
    }

    startClock() {
        for (var i = 0; i < this.minCount; i++) {
            this.addOneMin(this.main.shift());
        }

        var json = {"Min": this.mins, "FiveMin": this.fiveMinutes, "Hour": this.hoursAmount, "Main": this.main};
        console.log(json);
    }
}

function runClock() {
    new BallClock(45).startClock();
    new BallClockMins(30, 325).startClock();
}